File: services/audit_service.py
import datetime
import uuid
from services.firebase_service import FirebaseService

class AuditService:
    @staticmethod
    def log(action: str, performed_by: str, role: str, target_type: str, target_id: str, details: str = ""):
        log_id = f"log_{uuid.uuid4().hex[:12]}"
        timestamp = datetime.datetime.utcnow().isoformat() + "Z"
        if isinstance(details, dict):
            details_obj = details
        elif isinstance(details, str):
            details_obj = {"message": details} if details else {}
        else:
            details_obj = {"info": str(details)}

        log_data = {
            "logId": log_id,
            "action": action,
            "performedBy": performed_by,
            "role": role,
            "targetType": target_type,
            "targetId": target_id,
            "timestamp": timestamp,
            "details": details_obj
        }
        try:
            FirebaseService.set_document("auditLogs", log_id, log_data)
        except Exception as e:
            print(f"Failed to write audit log: {e}")
        return log_data

File: src/server/services/auditService.ts
import crypto from 'crypto';
import { FirebaseService } from './firebaseService.js';

export class AuditService {
  static async log(
    action: string,
    performedBy: string,
    role: string,
    targetType: string,
    targetId: string,
    details: any = ''
  ): Promise<Record<string, any>> {
    const logId = `log_${crypto.randomUUID().replace(/-/g, '').slice(0, 12)}`;
    const timestamp = new Date().toISOString();

    let detailsObj: Record<string, any>;
    if (typeof details === 'object' && details !== null) {
      detailsObj = details;
    } else if (typeof details === 'string') {
      detailsObj = details ? { message: details } : {};
    } else {
      detailsObj = { info: String(details) };
    }

    const logData = {
      logId,
      action,
      performedBy,
      role,
      targetType,
      targetId,
      timestamp,
      details: detailsObj,
      _id: logId
    };

    try {
      await FirebaseService.set_document('auditLogs', logId, logData);
    } catch (e) {
      console.warn('Failed to write audit log:', e);
    }
    return logData;
  }
}
